chain-tx-enclave/enclave-ra/ra-enclave/src/context.rs

```diff
@@ -0,0 +1,228 @@
+        let report_bytes: &[u8] = report.as_ref();
+
+        // Get quote and QE report from SP server
+        let quote_result = self.sp_ra_client.get_quote(report_bytes.to_vec())?;
+        let quote = quote_result.quote;
+        let qe_report_bytes = quote_result.qe_report;
+
+        // Verify QE report
+        let qe_report = Report::try_copy_from(&qe_report_bytes)
+            .ok_or_else(|| EnclaveRaContextError::InvalidQeReport)?;
+        verify_qe_report(&qe_report, &target_info)?;
+
+        // Get attestation report from SP server
+        self.sp_ra_client
+            .get_attestation_report(quote)
+            .map_err(Into::into)
+    }
+
+    /// Generates enclave report containing hash of public key of RA-TLS key-pair in user-data
+    fn get_report(
+        &self,
+        target_info: &Targetinfo,
+        public_key_hash: [u8; 32],
+    ) -> Result<Report, EnclaveRaContextError> {
+        let mut report_data = [0; 64];
+        (&mut report_data[0..32]).copy_from_slice(&public_key_hash);
+
+        Ok(Report::for_target(target_info, &report_data))
+    }
+
+    /// Creates new certificate
+    fn create_certificate(&self) -> Result<Certificate, EnclaveRaContextError> {
+        let certificate_params = self.create_certificate_params()?;
+
+        let private_key = PrivateKey(
+            certificate_params
+                .key_pair
+                .as_ref()
+                .ok_or_else(|| EnclaveRaContextError::MissingKeyPair)?
+                .serialize_der(),
+        );
+        let created = certificate_params.not_before;
+        let rustls_certificate =
+            RustlsCertificate(RcGenCertificate::from_params(certificate_params)?.serialize_der()?);
+
+        Ok(Certificate {
+            certificate: rustls_certificate,
+            created,
+            private_key,
+        })
+    }
+
+    /// Creates new certificate params
+    fn create_certificate_params(&self) -> Result<CertificateParams, EnclaveRaContextError> {
+        let mut certificate_params = CertificateParams::default();
+
+        certificate_params.alg = &PKCS_ECDSA_P256_SHA256;
+
+        let current_time = Utc::now();
+        certificate_params.not_before = current_time;
+        certificate_params.not_after = current_time + Duration::days(90);
+
+        certificate_params.subject_alt_names =
+            vec![SanType::Rfc822Name("security@crypto.com".to_string())];
+
+        let mut distinguished_name = DistinguishedName::new();
+        distinguished_name.push(DnType::OrganizationName, "Crypto.com");
+        distinguished_name.push(DnType::CommonName, "Crypto.com");
+        certificate_params.distinguished_name = distinguished_name;
+
+        certificate_params.is_ca = IsCa::SelfSignedOnly;
+
+        let key_pair = KeyPair::generate(&PKCS_ECDSA_P256_SHA256)?;
+
+        let attestation_report = self.get_attestation_report(sha256(key_pair.public_key_raw()))?;
```

Comment:
[This](https://arxiv.org/ftp/arxiv/papers/1801/1801.05863.pdf) paper mentions that we should put hash of public key.

Reply:
https://github.com/crypto-com/chain/pull/1311#discussion_r398476343

Also, as this is using that proxy SP thing rather than directly connecting to IAS over TLS (where it could verify the cert chain before connecting), it should verify its report as well -- otherwise, it'd bundle arbitrary payloads to the cert

Reply:
What things should be verify in attestation report? Also, what should be the behaviour when the verification fails? Currently, the certificate and attestation report is verified by client and they don't trust the enclave if any one of that verification fails.

Reply:
I thought the idea behind remote attestation was that the challenger needs some proof to trust the enclave. If the enclave is fine but it is connected to an untrusted SP, challenger will not trust the enclave.

Reply:
> What things should we verify in attestation report (in enclave)

it can do the same check that challenger is expected to do -- if there's something wrong, it'll fail fast instead of running along with arbitrary data.

> Also, what should be the behaviour when the verification fails?

Just fail / quit -- what's the point of running with wrong data?

>I thought the idea behind remote attestation was that the challenger needs some proof to trust the enclave. If the enclave is fine but it is connected to an untrusted and incorrect SP, challenger will not trust the enclave.

Yes, and it should still be that way (i.e. challenger still verifiers before connection). The point of server not bundling arbitrary data and doing some checks on them is an orthogonal concern.